### ISSUE TITLE
Installing on Windows: symlinking without elevated privileges 

### ISSUE THREAD

**Issue by zaps** [brigade/overcommit#400]
Hi, I'm trying to see if I can avoid having Windows users install Overcommit from an elevated command prompt, which is a requirement of the `mklink` command that is used to create symlinks.

In PR #373, @bsa01 modified `lib/overcommit/utils/file_utils.rb` to add the `/H` switch to `mklink`, which creates a hardlink instead of a softlink. This allows a non-admin user to create a link but it broke the `win32_symlink?` method in the file so the PR was closed.

I did an experiment with adding back the `/H` switch and it looks like it did not have an adverse affect on the operation of Overcommit. Granted I did not do extensive edge testing but for typical tasks it seems to work, installs and cleans up after itself. 

Here is an overview of the procedure I performed:

## Current behavior
1. Open a non-elevated Windows command or Git Bash 
2. `git clone http://my-repository`

     contents of `.git/hooks`:
     ```
     applypatch-msg.sample
     commit-msg.sample
     post-update.sample
     pre-applypatch.sample
     pre-commit.sample
     etc
```
3. `overcommit --install`

    contents of `.git/hooks:`
    ```
    applypatch-msg.sample
    commit-msg.sample
    overcommit-hook
    post-update.sample
    pre-applypatch.sample
    pre-commit.sample
    etc
    ```
   And none of the hooks work.

### No custom hooks
This is testing the installer with no custom hooks present in the repo (i.e. with just the default `*.sample` scripts)

1. Delete and reclone the repository
2. Patch the installed overcommit gem (latest 0.34.2) per #373
3. `overcommit --install`

    contents of `.git/hooks:`
    ```
    applypatch-msg.sample
    commit-msg
    commit-msg.sample
    overcommit-hook
    post-update.sample
    post-checkout
    post-commit
    pre-applypatch.sample
    pre-commit
    etc
    ```
   And the hooks work.
4. `overcommit --uninstall`

    contents of `.git/hooks:`
    ```
    applypatch-msg.sample
    commit-msg.sample
    post-update.sample
    pre-applypatch.sample
    pre-commit.sample
    etc
    ```
    The hooks are properly removed by the uninstaller.

### Yes custom hooks
One more time, this time simulating an existing custom `commit-msg` hook

1. Delete and reclone the repository
2. `cd .git\hooks`
3. `move commit-msg.sample commit.msg`
4. `cd ..\..`
5. `overcommit --install`
    
   Output
> Hook \path\to\.git\commit-msg already exists and was not installed by Overcommit

   The hooks are installed correctly and work.

6. `overcommit --uninstall`

    Output
> Successfully restored old hooks from \path\to\.git\hooks\old-hooks
    
    contents of `.git/hooks:`
    ```
    applypatch-msg.sample
    commit-msg
    post-update.sample
    pre-applypatch.sample
    pre-commit.sample
    etc
    ```

So it appears Overcommit is doing the right thing w/r/t installing and uninstalling the hooks with the `mklink /H` switch.

I understand why the original PR was closed due to not being able to parse hardlinks from a Windows directory listing. But if the installation and uninstallation work with `mklink /H` is it possible that the method could be bypassed if the detected OS is Windows?  

Am I crazy to suggest turning off **all** symlinking on Windows? :cloud: :fearful:

Thanks!

**Comment by sds** [comment]
Hey @zaps, thanks for your detailed investigation into this issue.

My short answer here is "I don't know" because I don't use Windows. @jawshooah did a tremendous job adding Windows support to Overcommit but I don't know the reasons why we preserved symlinking instead of avoiding it entirely. My guess is that it was largely to keep behavior as similar as possible between platforms.

However, as of 5bf372fd8ba543f90aeffee0e8d87bed4c1898e4 we explicitly stopped shipping the gem with symlinks in the `template-dir` directory, since newer versions of RubyGems will complain. What we didn't end up doing was augmenting `overcommit --install` to copy the hook files instead of symlinking when on Windows. This oversight is again mostly because I simply do not know the Windows world very well.

I would be happy to accept a pull request that fixes this issue by copying the hook files instead of symlinking them. Using symlinks is nice when you have the option since editing the `overcommit-hook` master hook will automatically update all the hooks that symlink to it, but it's not required for the hooks to work.

**Comment by zaps** [comment]
Ok cool thanks @sds for the lowdown -- I totally get the desire to keep feature parity between the OS's. 

It's been a while since I've Ruby-ed but I think this task might be small enough in scope that I could tackle it. As a first pass I'll work on switching `overcommit` `--install`/`--uninstall` to straight copies/deletes on Windows, since you've confirmed that not using symlinks won't break Overcommit.

A more robust enhancement would be to ostensibly replicate the functionality of symlinks, like how you mentioned; a change to `overcommit-hook` would trigger a "copy scripts" operation on Windows, keeping everything in sync. That one might be a bit trickier so I'll hit the easy one first :smirk: 

Thanks again

**Comment by sds** [comment]
Implemented in #403
